kernel/dw1000.c

```diff
@@ -1247,36 +1249,26 @@ static void dw1000_ptp_worker(struct work_struct *work)
 static int dw1000_ptp_adjfreq(struct ptp_clock_info *ptp, int32_t delta)
 {
 	struct dw1000 *dw = container_of(ptp, struct dw1000, ptp.info);
-	uint64_t delta_mult;
-	uint32_t mult;
-	bool negate;
-
-	/* Split out sign to allow for unsigned division */
-	if (delta < 0) {
-		delta = -delta;
-		negate = true;
-	} else {
-		negate = false;
-	}
+	struct hires_counter *tc = &dw->ptp.tc;
+	u64 mult;
+
+	/* Split out sign to allow for unsigned multiplication */
+	mult = (delta < 0) ? -delta : delta;
 
 	/* Calculate multiplier value */
-	delta_mult = delta;
-	delta_mult *= DW1000_CYCLECOUNTER_MULT;
-	delta_mult = div_u64(delta_mult, 1000000000UL); /* ppb */
-	mult = (negate ? (DW1000_CYCLECOUNTER_MULT - delta_mult) :
-		(DW1000_CYCLECOUNTER_MULT + delta_mult));
-	dev_dbg(dw->dev, "adjust frequency %+d ppb: multiplier %u->%u\n",
-		(negate ? -delta : delta), DW1000_CYCLECOUNTER_MULT, mult);
-
-	/* Read timecounter to establish a baseline point at which the
-	 * frequency changes (i.e. to synchronise the cycle and
-	 * nanosecond counts), then adjust the multiplier.
-	 */
+	mult *= DW1000_CYCLECOUNTER_MULT / 1000000000ULL;
```

Comment:
You've probably checked this already, but just in case: does this calculation avoid both potential overflow and potential inaccuracy due to the precalculated division?

Reply:
The result is 288692283, which is certainly accurate enough. It allows delta to have range of -1E10 .. 1E10